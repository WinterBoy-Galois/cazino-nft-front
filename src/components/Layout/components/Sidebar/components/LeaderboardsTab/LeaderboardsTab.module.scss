@import '../../../../../../styles/shared';

.table {
  height: calc(100% - #{$sidebar-time-select-height});
  padding-bottom: 1rem;
}

.time-select {
  margin-bottom: 1.5rem;

  @include media-breakpoint-up(md) {
    margin-bottom: 2rem;
  }
}
