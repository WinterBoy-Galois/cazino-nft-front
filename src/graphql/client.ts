import { ApolloClient, InMemoryCache, from, split } from '@apollo/client';
import { WebSocketLink } from '@apollo/client/link/ws';
import { TokenRefreshLink } from 'apollo-link-token-refresh';
import possibleTypes from './possibleTypes.json';
import { getMainDefinition } from '@apollo/client/utilities';
import { HttpLink, defaultDataIdFromObject } from '@apollo/client';
import jwtDecode from 'jwt-decode';
import { getEpoch } from '../common/util/date.util';
import { setContext } from '@apollo/client/link/context';
import { AuthType } from '../state/models/auth.model';
import { appConfig } from '../common/config';

const cache = new InMemoryCache({
  dataIdFromObject(responseObject) {
    switch (responseObject.__typename) {
      default:
        return defaultDataIdFromObject(responseObject);
    }
  },
  addTypename: true,
  possibleTypes,
});

let wsLink: WebSocketLink;

const getApolloClient = (
  onAccessTokenRefresh: (accessToken: string) => void,
  onSignOut: () => void,
  authType: AuthType,
  accessToken?: string
) => {
  cache.reset();

  if (wsLink) {
    (wsLink as any).subscriptionClient.close();
  }

  wsLink = new WebSocketLink({
    uri: `${appConfig.apiBasePathWS}/graphql`,
    options: {
      reconnect: true,
      connectionParams: {
        authToken: accessToken ?? undefined,
      },
    },
  });

  const httpLink = new HttpLink({
    uri: `${appConfig.apiBasePath}/graphql`,
    credentials: 'include',
  });

  const authLink = setContext((_, { headers }) => {
    if (accessToken) {
      return {
        headers: {
          ...headers,
          authorization: `Bearer ${accessToken}`,
        },
      };
    }

    return headers;
  });

  const tokenLink = new TokenRefreshLink({
    accessTokenField: 'accessToken',
    isTokenValidOrUndefined: () => {
      if (accessToken) {
        const { exp } = jwtDecode(accessToken);

        if (exp < getEpoch()) {
          return false;
        }
      } else if (authType === 'SIGNED_IN') {
        return false;
      }

      return true;
    },
    fetchAccessToken: () => {
      return fetch(`${appConfig.apiBasePath}/refresh_tokens`, {
        method: 'POST',
        credentials: 'include',
      });
    },
    handleFetch: accessToken => onAccessTokenRefresh(accessToken),
    handleError: () => onSignOut(),
  });

  const link = split(
    ({ query }) => {
      const definition = getMainDefinition(query);
      return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
    },
    wsLink,
    authLink.concat(httpLink)
  );

  return new ApolloClient({
    cache,
    link: from([tokenLink, link]),
  });
};

export default getApolloClient;
