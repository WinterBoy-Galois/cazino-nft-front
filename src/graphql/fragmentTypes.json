{
  "__schema": {
    "types": [
      {
        "kind": "UNION",
        "name": "PublicUserOrError",
        "possibleTypes": [{ "name": "PublicUser" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "LoginResultOrError",
        "possibleTypes": [{ "name": "LoginResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "UserOrError",
        "possibleTypes": [{ "name": "User" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "SignedInTokenOrError",
        "possibleTypes": [{ "name": "LoginResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "BetResultOrError",
        "possibleTypes": [{ "name": "BetResult" }, { "name": "GenericErrorArray" }]
      },
      {
        "kind": "UNION",
        "name": "BooleanOrError",
        "possibleTypes": [{ "name": "GenericBoolean" }, { "name": "GenericErrorArray" }]
      }
    ]
  }
}
