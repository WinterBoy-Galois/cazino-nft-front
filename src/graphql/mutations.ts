import gql from 'graphql-tag';

export const CHANGE_SERVER_SEED = gql`
  mutation ChangeServerSeed {
    changeServerSeed {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ACTIVATE_ACCOUNT = gql`
  mutation ActivateAccount($code: String!) {
    activateAccount(code: $code) {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_IN = gql`
  mutation SignIn($email: String, $password: String, $remember: Boolean) {
    signIn(email: $email, password: $password, remember: $remember) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          email
          isActivated
          depositAddress
          refCode
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_UP = gql`
  mutation SignUp(
    $email: String
    $password: String
    $username: String
    $token: String
    $ref: String
  ) {
    registerUser(
      username: $username
      email: $email
      password: $password
      captcha: $token
      ref: $ref
    ) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          depositAddress
          refCode
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SIGN_OUT = gql`
  mutation SignOut {
    signOut
  }
`;

export const RESEND_ACTIVATION_CODE = gql`
  mutation ResendActivationCode {
    resendActivationCode {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const RESET_PASSWORD = gql`
  mutation ResetPassword($token: String!, $newPassword: String!) {
    resetPassword(uuid: $token, newPassword: $newPassword) {
      ... on LoginResult {
        accessToken
        user {
          id
          username
          avatarUrl
          balance
          depositAddress
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const RECOVER_PASSWORD = gql`
  mutation RecoverPassword($email: String!) {
    forgotPassword(email: $email) {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_PASSWORD = gql`
  mutation UpdatePassword($oldPassword: String!, $newPassword: String!) {
    modifyPassword(oldPassword: $oldPassword, newPassword: $newPassword) {
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_PREFERENCES = gql`
  mutation UpdatePreferences($hideUsername: Boolean!, $hideProfit: Boolean!, $hideWager: Boolean!) {
    modifyPreferences(
      hideUsername: $hideUsername
      hideTotalProfit: $hideProfit
      hideTotalWager: $hideWager
    ) {
      ... on User {
        id
        hideUsername
        hideTotalProfit
        hideTotalWager
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const UPDATE_AVATAR = gql`
  mutation updateAvatar($index: Int!) {
    modifyAvatar(index: $index) {
      ... on User {
        id
        avatarUrl
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_DICE = gql`
  mutation makeBetDice($betAmount: Float!, $over: Boolean!, $target: Float!) {
    makeBetDice(betAmount: $betAmount, over: $over, target: $target) {
      ... on BetResultDice {
        id
        result
        lucky
        multiplier
        profit
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_CLAMS = gql`
  mutation makeBetClams($betAmount: Float!, $selection: [Int]) {
    makeBetClams(betAmount: $betAmount, selection: $selection) {
      ... on BetResultClams {
        id
        result
        lucky
        multiplier
        profit
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_GOALS = gql`
  mutation makeBetGoals($betAmount: Float!, $difficulty: GoalsDifficulty!) {
    makeBetGoals(betAmount: $betAmount, difficulty: $difficulty) {
      ... on GoalsGameSetup {
        session {
          betId
          betAmount
          currentStep
          selections {
            step
            luckySpots
            selected
          }
          difficulty
          profits {
            step
            multiplier
            profit
          }
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ADVANCE_GOALS = gql`
  mutation advanceGoals($betId: ID!, $selection: Int!) {
    advanceGoals(betId: $betId, selection: $selection) {
      ... on GoalsStep {
        id
        totalProfit {
          multiplier
          profit
        }
        nextProfit {
          multiplier
          profit
        }
        nextStep
        profitCut
        allowNext
      }
      ... on GoalsComplete {
        id
        lucky
        profit {
          multiplier
          profit
        }
        balance
        profitCut
        result {
          step
          luckySpots
          selected
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CASH_OUT_GOALS = gql`
  mutation cashoutGoals($betId: ID!) {
    cashoutGoals(betId: $betId) {
      ... on GoalsComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        profitCut
        result {
          step
          luckySpots
          selected
        }
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const WITHDRAW = gql`
  mutation withdraw($amount: Float!, $address: String!) {
    withdraw(amount: $amount, address: $address) {
      ... on GenericBoolean {
        result
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CLAIM_FAUCET = gql`
  mutation claimFaucet {
    claimFaucet {
      ... on User {
        balance
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const MAKE_BET_MINES = gql`
  mutation makeBetMines($betAmount: Float!, $mines: Int!) {
    makeBetMines(betAmount: $betAmount, mines: $mines) {
      __typename
      ... on MinesGameSetup {
        session {
          ... on MinesGameSessionSetup {
            betId
            betAmount
            mines
            open
            totalProfit {
              multiplier
              profit
            }
            nextProfit {
              multiplier
              profit
            }
            allowNext
            profitCut
          }
        }
        balance
        maxProfit
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const ADVANCE_MINES = gql`
  mutation advanceMines($betId: ID!, $selection: Int!) {
    advanceMines(betId: $betId, selection: $selection) {
      ... on MinesStep {
        id
        totalProfit {
          multiplier
          profit
        }
        nextProfit {
          multiplier
          profit
        }
        profitCut
        allowNext
        result
      }
      ... on MinesComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        minePositions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CASH_OUT_MINES = gql`
  mutation cashoutMines($betId: ID!) {
    cashoutMines(betId: $betId) {
      ... on MinesComplete {
        id
        lucky
        balance
        profit {
          multiplier
          profit
        }
        minePositions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const CLAIM_BONUS = gql`
  mutation claimBonus($bonusId: ID!) {
    claimBonus(bonusId: $bonusId) {
      ... on User {
        balance
      }
      ... on GenericErrorArray {
        errors {
          code
          message
        }
      }
    }
  }
`;

export const CLAIM_COMMISSION = gql`
  mutation claimCommissions {
    claimCommissions {
      ... on User {
        id
        username
        email
        balance
        hideUsername
        hideTotalProfit
        hideTotalWager
        isActivated
        avatarUrl
        totalWager
        totalProfit
        mostPlayed
        totalBets
        luckyBets
        depositAddress
        refCode
        refCommissions
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;
