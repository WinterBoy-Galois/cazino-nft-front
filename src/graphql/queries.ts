import gql from 'graphql-tag';

export const RECENT_BETS = gql`
  {
    recentBets {
      allBets {
        id
        time
        userid
        username
        gameid
        bet
        profit
        multiplier
      }
      myBets {
        id
        time
        userid
        username
        gameid
        bet
        profit
        multiplier
      }
    }
  }
`;

export const LEADERBOARDS = gql`
  {
    leaderboards {
      daily {
        username
        userid
        wager
        bonus
      }
      weekly {
        username
        userid
        wager
        bonus
      }
      monthly {
        username
        userid
        wager
        bonus
      }
    }
  }
`;

export const USER_INFO = gql`
  query UserInfo($userId: ID) {
    userInfo(id: $userId) {
      __typename
      ... on PublicUser {
        id
        username
        avatarUrl
        totalWager
        totalProfit
        mostPlayed
        totalBets
        luckyBets
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const USER_INFO_AVATAR_URL = gql`
  query UserInfoAvatarUrl($userId: ID) {
    userInfo(id: $userId) {
      __typename
      ... on PublicUser {
        id
        avatarUrl
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const BET_DETAILS = gql`
  query BetDetails($betId: ID) {
    betDetails(id: $betId) {
      ... on BetDetails {
        bet
        profit
        profitCut
        multiplier
        gameResult {
          ... on DiceResult {
            target
            over
            winChance
            resultFloat
          }
          ... on MinesResult {
            mineCount
            minePositions
            open
          }
          ... on GoalsResult {
            difficulty
            selections {
              step
              luckySpots
              selected
            }
          }
          ... on ClamsResult {
            selection
            resultInteger
          }
        }
      }
    }
  }
`;

export const BET_DETAILS_SERVER_SEED = gql`
  query BetDetailsServerSeed($betId: ID) {
    betDetails(id: $betId) {
      ... on BetDetails {
        seedDetails {
          __typename
          ... on SeedDetailsOther {
            serverSeedHash
          }
          ... on SeedDetailsOwn {
            activeGames
            serverSeedHash
            serverSeed
            clientSeed
            nonce
            results
            verificationUrl
          }
        }
      }
    }
  }
`;

export const ME = gql`
  query Me {
    me {
      id
      username
      avatarUrl
      balance
      email
      isActivated
      depositAddress
      refCommissions
      refCode
    }
  }
`;

export const AFF_STATS = gql`
  query affStats {
    affStats {
      bets
      refs
      wager
      commissions
    }
  }
`;

export const ME_STATISTICS_PREFERENCES = gql`
  query Me {
    me {
      id
      totalWager
      totalProfit
      mostPlayed
      totalBets
      luckyBets
      hideUsername
      hideTotalProfit
      hideTotalWager
    }
  }
`;

export const SETUP_CASHIER = gql`
  query SetupCashier {
    setupCashier {
      networkFee
      depositConfirmations
      minWithdraw
    }
    me {
      id
      depositAddress
    }
  }
`;

export const DEPOSITS = gql`
  query Deposits($page: Int, $limit: Int) {
    transactionsDeposit(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        status
        time
        hash
        amount
      }
    }
  }
`;

export const SETUP_DICE = gql`
  query SetupDice {
    setupDice {
      ... on DiceGameSetup {
        minProbability
        maxProbability
        he
      }
    }
  }
`;

export const SETUP_CLAMS = gql`
  query setupClams {
    setupClams {
      ... on ClamsGameSetup {
        he
      }
    }
  }
`;

export const SETUP_GOAL = gql`
  query setupGoals {
    setupGoals {
      ... on GoalsGameSetup {
        session {
          betId
          betAmount
          currentStep
          selections {
            step
            luckySpots
            selected
          }
          difficulty
          profits {
            step
            multiplier
            profit
          }
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
        maxProfit
      }
    }
  }
`;

export const TRANSACTION_BETS = gql`
  query Bets($page: Int, $limit: Int) {
    transactionsBet(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        time
        game
        amount
        multiplier
        profit
      }
    }
  }
`;

export const TRANSACTION_AFFILIATES = gql`
  query AffiliateClaims($page: Int, $limit: Int) {
    transactionsAffiliate(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        claimedAt
        amount
      }
    }
  }
`;

export const TRANSACTION_WITHDRAWALS = gql`
  query transactionsWithdraw($page: Int, $limit: Int) {
    transactionsWithdraw(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        status
        time
        hash
        address
        amount
        fee
      }
    }
  }
`;

export const TRANSACTION_BONUSES = gql`
  query transactionsBonus($page: Int, $limit: Int) {
    transactionsBonus(page: $page, limit: $limit) {
      page
      limit
      total
      items {
        id
        givenAt
        claimedAt
        expiresAt
        amount
        position
        type
        wager
      }
    }
  }
`;

export const FAUCET_INFO = gql`
  query faucetInfo {
    faucetInfo {
      ... on FaucetInfo {
        amount
        every
        canClaim
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const SETUP_MINES = gql`
  query setupMines {
    setupMines {
      ... on MinesGameSetup {
        session {
          betId
          betAmount
          mines
          open
          totalProfit {
            multiplier
            profit
          }
          nextProfit {
            multiplier
            profit
          }
          allowNext
          profitCut
        }
        balance
        maxProfit
      }
      ... on GenericErrorArray {
        errors {
          source
          code
          message
          args
        }
      }
    }
  }
`;

export const BONUSCLAIMS = gql`
  query bonusClaims {
    bonusClaims {
      ... on BonusClaim {
        id
        givenAt
        claimedAt
        expiresAt
        amount
        position
        type
        wager
      }
    }
  }
`;
