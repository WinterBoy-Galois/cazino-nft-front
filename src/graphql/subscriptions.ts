import gql from 'graphql-tag';

export const LEADERBOARDS_SUBSCRIPTION = gql`
  subscription leaderboardChanged {
    leaderboardChanged {
      daily {
        username
        userid
        wager
        bonus
      }
      weekly {
        username
        userid
        wager
        bonus
      }
      monthly {
        username
        userid
        wager
        bonus
      }
    }
  }
`;

export const BET_ADDED = gql`
  subscription onBetAdded {
    betAdded {
      id
      time
      userid
      username
      gameid
      bet
      profit
      multiplier
    }
  }
`;

export const BALANCE = gql`
  subscription balance {
    balance {
      user {
        id
        balance
      }
      amount
      event
    }
  }
`;
