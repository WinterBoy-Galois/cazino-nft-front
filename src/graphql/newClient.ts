import {
  ApolloClient,
  defaultDataIdFromObject,
  from,
  fromPromise,
  HttpLink,
  InMemoryCache,
  Operation,
  split,
} from '@apollo/client';
import possibleTypes from './possibleTypes.json';
import { WebSocketLink } from '@apollo/client/link/ws';
import { appConfig } from '../common/config';
import { setContext } from '@apollo/client/link/context';
import { getMainDefinition } from '@apollo/client/utilities';
import { onError } from '@apollo/client/link/error';
import { error } from '../components/Toast';
import { GraphQLError } from 'graphql';
import { RetryLink } from '@apollo/client/link/retry';
import { getAccessToken, setAccessToken } from '../user/UserProvider';

const cache = new InMemoryCache({
  dataIdFromObject(responseObject) {
    switch (responseObject.__typename) {
      default:
        return defaultDataIdFromObject(responseObject);
    }
  },
  addTypename: true,
  possibleTypes,
});

const wsLink = new WebSocketLink({
  uri: `${appConfig.apiBasePathWS}/graphql`,
  options: {
    reconnect: true,
    connectionParams: {
      authToken: getAccessToken() ?? undefined,
    },
  },
});

const httpLink = new HttpLink({
  uri: `${appConfig.apiBasePath}/graphql`,
  credentials: 'include',
});

const authLink = setContext((_, { headers }) => {
  const accessToken = getAccessToken();
  if (accessToken) {
    return {
      headers: {
        ...headers,
        authorization: `Bearer ${accessToken}`,
      },
    };
  }

  return headers;
});

const link = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return definition.kind === 'OperationDefinition' && definition.operation === 'subscription';
  },
  wsLink,
  authLink.concat(httpLink)
);

let isRefreshing = false;
let pendingRequests: any[] = [];

const resolvePendingRequests = () => {
  pendingRequests.map(callback => {
    return callback();
  });
  pendingRequests = [];
};

let request: any;

export const getNewToken = () => {
  if (!request) {
    request = fetch(`${appConfig.apiBasePath}/refresh_tokens`, {
      method: 'POST',
      credentials: 'include',
    });
  }
  return request
    .then((result: any) => {
      if (!result.ok) {
        throw new GraphQLError(`${result.status}: ${result.statusText}`);
      }
      return result.json();
    })
    .finally(() => {
      request = null;
    });
};

const newOperation = (operation: Operation) => {
  const accessToken = getAccessToken();
  if (accessToken) {
    operation.setContext({
      headers: {
        ...operation.getContext().headers,
        authorization: `Bearer ${accessToken}`,
      },
    });
  }
  return operation;
};

const retryLink = new RetryLink({
  attempts: {
    retryIf: error => !!error,
    max: 5,
  },
  delay: () => 1000,
});

export const useApolloClient = (logout: any) => {
  const errorLink = onError(({ graphQLErrors, operation, forward, networkError, response }) => {
    if (graphQLErrors) {
      for (const err of graphQLErrors) {
        switch (err?.extensions?.code) {
          case 'FORBIDDEN':
            let forward$;

            if (!isRefreshing && operation.operationName !== 'SignOut') {
              isRefreshing = true;
              forward$ = fromPromise(
                getNewToken()
                  .then(({ accessToken }: any) => {
                    setAccessToken(accessToken);
                    resolvePendingRequests();
                    return accessToken;
                  })
                  .catch(() => {
                    logout?.();
                    return forward(operation);
                  })
                  .finally(() => {
                    isRefreshing = false;
                  })
              ).filter(value => Boolean(value));
            } else {
              forward$ = fromPromise(
                new Promise(resolve => {
                  pendingRequests.push(() => resolve());
                })
              );
            }
            return forward$.flatMap(() => {
              return forward(newOperation(operation));
            });
        }
      }
      graphQLErrors.map(({ message, locations, path }) =>
        // eslint-disable-next-line no-console
        console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`)
      );
      if (networkError) {
        if (networkError && 'statusCode' in networkError && networkError.statusCode === 503) {
          error(networkError.message);
          if (response) {
            response.errors = undefined;
          }
        }
        // eslint-disable-next-line no-console
        console.log('Error--------------------', networkError);
      }
    }
  });

  const apolloLink = from([errorLink, retryLink, link]);

  return new ApolloClient({
    cache,
    link: apolloLink,
    queryDeduplication: false,
    connectToDevTools: process.env.NODE_ENV === 'development',
  });
};
