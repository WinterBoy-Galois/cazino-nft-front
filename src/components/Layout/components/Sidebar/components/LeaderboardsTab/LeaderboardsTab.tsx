import React, { useState, useEffect } from 'react';
import styles from './LeaderboardsTab.module.scss';
import { useTranslation } from 'react-i18next';
import SlideSelect from '../../../../../SlideSelect';
import LeaderboardTable from '../../../../../LeaderboardTable';
import { useQuery } from '@apollo/client';
import { LEADERBOARDS_SUBSCRIPTION } from '../../../../../../graphql/subscriptions';
import { LEADERBOARDS } from '../../../../../../graphql/queries';
import { useLocation, useNavigate } from '@reach/router';

const LeaderboardsTab: React.FC = () => {
  const [selectedTime, setSelectedTime] = useState<TimeAggregation>('daily');
  const { t } = useTranslation(['sidebar']);
  const { loading, error, data, subscribeToMore } = useQuery(LEADERBOARDS);
  const { pathname } = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    return subscribeToMore({
      document: LEADERBOARDS_SUBSCRIPTION,
      updateQuery: (prev: any, { subscriptionData }: any) =>
        subscriptionData.data
          ? { ...prev, leaderboards: { ...(subscriptionData.data as any).leaderboardChanged } }
          : prev,
    });
  }, [subscribeToMore]);

  return (
    <>
      <div className={styles['time-select']}>
        <SlideSelect
          selectItems={[
            { label: t('leaderboards.times.daily'), onClick: () => setSelectedTime('daily') },
            { label: t('leaderboards.times.weekly'), onClick: () => setSelectedTime('weekly') },
            { label: t('leaderboards.times.monthly'), onClick: () => setSelectedTime('monthly') },
          ]}
        />
      </div>
      <div className={styles.table}>
        <LeaderboardTable
          leaderboard={data ? data.leaderboards[selectedTime] : []}
          isLoading={loading}
          error={error ? true : false}
          signInUserId="15"
          onUsernameClicked={userId =>
            navigate(`${pathname}?dialog=user-info`, { state: { userId } })
          }
        />
      </div>
    </>
  );
};

export default LeaderboardsTab;

type TimeAggregation = 'daily' | 'weekly' | 'monthly';
