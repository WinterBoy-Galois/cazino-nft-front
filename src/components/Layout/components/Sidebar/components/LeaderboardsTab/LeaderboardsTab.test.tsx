import React from 'react';
import { render } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import LeaderboardsTab from './LeaderboardsTab';
import { LEADERBOARDS_SUBSCRIPTION } from '../../../../../../graphql/subscriptions';
import { LEADERBOARDS } from '../../../../../../graphql/queries';
import { LocationProvider } from '@reach/router';

const mocks = [
  {
    request: {
      query: LEADERBOARDS,
    },
    result: {
      data: { leaderboards: { daily: [], weekly: [], monthly: [] } },
    },
  },
  {
    request: {
      query: LEADERBOARDS_SUBSCRIPTION,
    },
    result: {
      data: { leaderboardChanged: { daily: [], weekly: [], monthly: [] } },
    },
  },
];

describe('LeaderboardsTab', () => {
  it('should match snapshot', async () => {
    // Arrange

    // Act
    const container = render(
      <LocationProvider>
        <MockedProvider mocks={mocks} addTypename={false}>
          <LeaderboardsTab />
        </MockedProvider>
      </LocationProvider>
    );

    // Assert
    expect(container).toMatchSnapshot();
  });
});
