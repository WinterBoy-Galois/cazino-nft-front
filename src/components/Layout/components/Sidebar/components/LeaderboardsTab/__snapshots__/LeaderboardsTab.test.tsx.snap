// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardsTab should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="time-select"
      >
        <div
          class="container"
        >
          <span
            class="tab tab--selected"
            style="transform: translateX(0%); width: 33.333333333333336%;"
          />
          <span
            class="tab"
            style="width: 33.333333333333336%;"
          >
            Daily
          </span>
          <span
            class="tab"
            style="width: 33.333333333333336%;"
          >
            Weekly
          </span>
          <span
            class="tab"
            style="width: 33.333333333333336%;"
          >
            Monthly
          </span>
        </div>
      </div>
      <div
        class="table"
      >
        <div
          class="leaderboard-table__wrapper"
        >
          <table
            class="leaderboard-table"
          >
            <thead
              class="leaderboard-table__header"
            >
              <tr>
                <th>
                  #
                </th>
                <th>
                  User
                </th>
                <th>
                  Wagered
                </th>
                <th>
                  Bonus
                </th>
              </tr>
            </thead>
            <tbody
              class="leaderboard-table__body"
            />
          </table>
          <div
            class=" container"
          >
            <div
              class="spinner"
            >
              <span
                class="spinner spinner--white "
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="time-select"
    >
      <div
        class="container"
      >
        <span
          class="tab tab--selected"
          style="transform: translateX(0%); width: 33.333333333333336%;"
        />
        <span
          class="tab"
          style="width: 33.333333333333336%;"
        >
          Daily
        </span>
        <span
          class="tab"
          style="width: 33.333333333333336%;"
        >
          Weekly
        </span>
        <span
          class="tab"
          style="width: 33.333333333333336%;"
        >
          Monthly
        </span>
      </div>
    </div>
    <div
      class="table"
    >
      <div
        class="leaderboard-table__wrapper"
      >
        <table
          class="leaderboard-table"
        >
          <thead
            class="leaderboard-table__header"
          >
            <tr>
              <th>
                #
              </th>
              <th>
                User
              </th>
              <th>
                Wagered
              </th>
              <th>
                Bonus
              </th>
            </tr>
          </thead>
          <tbody
            class="leaderboard-table__body"
          />
        </table>
        <div
          class=" container"
        >
          <div
            class="spinner"
          >
            <span
              class="spinner spinner--white "
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
